### PR TITLE
perf: use svg sprite for icons

### PR DESCRIPTION
**Pull request by G3root**
## What does this PR do?

Rendering icons as JSX components is expensive and significantly increases the size of your bundle. This PR aims to address that by replacing them with SVG sprite icons, which are cacheable and faster than the JS based solution.

here's a explanation why we shouldn't import SVG icons as JSX https://twitter.com/_developit/status/1382838799420514317?lang=en

> [!NOTE]  
> Since this change is one-to-one with the previous implementation, it might still introduce some regressions, so it needs to be tested thoroughly

### here's how it's looks like after the changes
[Screencast 2024-08-08 18:45:30.webm](https://github.com/user-attachments/assets/cdcc2249-398a-4bec-925f-e84545edab56)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

### CODE REVIEW

File: packages/ui/scripts/generate-icons.mjs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import glob from "fast-glob";
+import fs from "fs-extra";
+import path from "node:path";
+
```
Comment (G3root) [inline packages/ui/scripts/generate-icons.mjs:4]: script to import svg from lucide icons

File: packages/ui/components/icon/icon-list.mjs [hunk 2]
```diff
@@ -0,0 +1,136 @@
+// icon name from lucide-react
+// this list is extracted based on what icons are currently used
+export const lucideIconList = new Set([
+  "x",
```
Comment (G3root) [inline packages/ui/components/icon/icon-list.mjs:4]: list of icon names used throughout the app

File: apps/web/public/icons/sprite.svg [hunk 3]
```diff
@@ -0,0 +1,641 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (G3root) [inline apps/web/public/icons/sprite.svg:1]: sprite sheet generated from `build-icons` command

File: packages/ui/package.json [hunk 4]
```diff
@@ -20,7 +20,8 @@
     "type-check": "tsc --pretty --noEmit",
     "type-check:ci": "tsc-absolute --pretty --noEmit",
     "lint:fix": "eslint . --fix",
-    "lint:report": "eslint . --format json --output-file ../../lint-results/ui.json"
+    "lint:report": "eslint . --format json --output-file ../../lint-results/ui.json",
+    "build-icons": "node ./scripts/build-icons.mjs"
```
Comment (G3root) [inline packages/ui/package.json:24]: command to generate the sprite sheet when new icons are added to the list

File: packages/ui/components/data-table/DataTableFilter.tsx [hunk 5]
```diff
@@ -23,7 +22,8 @@ interface DataTableFilter<TData, TValue> {
   options: {
     label: string;
     value: string;
-    icon?: LucideIcon;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    icon?: any;
```
Comment (PeerRich) [inline packages/ui/components/data-table/DataTableFilter.tsx:26]: do we have a type here?

File: packages/ui/components/data-table/DataTableToolbar.tsx [hunk 6]
```diff
@@ -17,7 +16,8 @@ export type FilterableItems = {
   options: {
     label: string;
     value: string;
-    icon?: LucideIcon;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    icon?: any;
```
Comment (PeerRich) [inline packages/ui/components/data-table/DataTableToolbar.tsx:20]: same here

File: packages/ui/components/icon/icon-list.mjs (hunk 2 above)
Comment (PeerRich) [reply packages/ui/components/icon/icon-list.mjs:4]: if we want to add a new icon, we have to add it here?

**Review by PeerRich — COMMENTED** [review]
this looks all fun and games but what is the actual performance difference?

**Comment by PeerRich** [comment]
requesting @zomars review here as he worked on the Icon component the most

File: packages/ui/components/data-table/DataTableFilter.tsx (hunk 5 above)
Comment (zomars) [reply packages/ui/components/data-table/DataTableFilter.tsx:26]: Yeah. This should be either a ReactNode or IconName

File: packages/ui/components/icon/README.md
Comment (zomars) [inline packages/ui/components/icon/README.md:1]: I appreciate the README 🙏🏽

File: packages/ui/components/top-banner/TopBanner.tsx [hunk 7]
```diff
@@ -21,8 +20,7 @@ const variantClassName = {
 
 const defaultIconProps = {
   className: "text-emphasis h-4 w-4 stroke-[2.5px]",
-  "aria-hidden": "true",
```
Comment (zomars) [inline packages/ui/components/top-banner/TopBanner.tsx:24]: I think we want icons not being read by screen readers by default.

**Comment by zomars** [comment]
Got some missing icons 
image

File: packages/ui/package.json (hunk 4 above)
Comment (zomars) [inline packages/ui/package.json:24]:
```suggestion
    "build:icons": "node ./scripts/build-icons.mjs"
```

To match the READMEs and comments

File: packages/ui/components/icon/icon-list.mjs
Comment (zomars) [inline packages/ui/components/icon/icon-list.mjs:1]: Sorted alphabetically

**Review by zomars — APPROVED** [review]

**Comment by sean-brydon** [comment]
CleanShot 2024-08-09 at 11 22 23
This commit caused password inputs and any inputs with addon-leading/suffix to break massively

**Comment by PeerRich** [comment]
> this looks all fun and games but what is the actual performance difference?


do we know the performance difference @G3root ??

**Comment by G3root** [comment]
@PeerRich  here's the breakdown 
- The gzipped final bundle size went from `8.29 MB` to `5.95 MB` according to `next-bundle-analyzer`.
- The first load JS for all pages went from `233 KB` to `195 KB`.
- While testing `/apps/installed/analytics` in an incognito tab, the JS bundle size went from `6.26 MB` to `5.95 MB`, and the load time decreased from `2.26 s` to `1.80 s`.


### before 

Screenshot 2024-08-09 at 20-19-38 @calcom_web 9 Aug 2024 at 20 12

Screenshot 2024-08-09 20:21:49


Screenshot 2024-08-09 20:24:02


### after

Screenshot 2024-08-09 at 18-54-50 @calcom_web 9 Aug 2024 at 18 52

Screenshot 2024-08-09 19:01:30


Screenshot 2024-08-09 19:04:35

**Comment by G3root** [comment]
Here’s an example of before and after. Before, when refreshing the page, the icons would flicker and show a skeleton. After, the icons no longer flash during refresh. there are some frame drops in the video due to my screen recorder

### before

[Screencast 2024-08-09 20:24:56.webm](https://github.com/user-attachments/assets/b3938117-baf5-49d8-9869-22ee9f48d622)

### after

[Screencast 2024-08-09 20:41:25.webm](https://github.com/user-attachments/assets/0ad1849e-6a5b-47e6-8eb6-4efd38a6ba2e)

**Comment by Ryukemeister** [comment]
@G3root what if we want to use the `Icon` component outside of the web app, won't that cause the icons to break in this case?

**Comment by zomars** [comment]
Just a heads up there's a fix for SVG and CORS in here https://github.com/calcom/cal.com/pull/16244

**Pull request merged**